core/drivers/counter/stm32_stgen.c

```diff
@@ -0,0 +1,270 @@
+ * No arguments.
+ */
+#define STM32_SIP_SMC_STGEN_SET_RATE		0x82000000
+
+/*
+ * struct stgen_pdata - STGEN instance
+ * @stgen_clock: STGEN subsystem clock
+ * @stgen_bus: STGEN interface bus clock
+ * @calendar: RTC calendar data at suspend time
+ * @cnt_h: STGEN high counter value at suspend time
+ * @cnt_l: STGEN low counter value at suspend time
+ * @base: STGEN IOMEM base address
+ */
+struct stgen_pdata {
+	struct clk *stgen_clock;
+	struct clk *bus_clock;
+	struct optee_rtc_time *calendar;
+	uint32_t cnt_h;
+	uint32_t cnt_l;
+	vaddr_t base;
+};
+
+static struct stgen_pdata stgen_d;
+
+static void set_counter_value(uint64_t counter)
+{
+	io_write64(stgen_d.base + STGENC_CNTCVL, counter);
+}
+
+uint64_t stm32_stgen_get_counter_value(void)
+{
+	return io_read64(stgen_d.base + STGENC_CNTCVL);
+}
+
+/*
+ * This function invokes EL3 monitor to update CNTFRQ with STGEN
+ * frequency. Updating the CP15 register can only be done by the
+ * highest EL on Armv8-A.
+ */
+static TEE_Result stm32mp_stgen_smc_config(void)
+{
+	/*
+	 * STM32_SIP_SMC_STGEN_SET_RATE call API
+	 *
+	 * Argument a0: (input) SMCC ID
+	 *		(output) status return code
+	 */
+	struct thread_smc_args stgen_conf_args = {
+		.a0 = STM32_SIP_SMC_STGEN_SET_RATE
+	};
+
+	thread_smccc(&stgen_conf_args);
+
+	if (stgen_conf_args.a0) {
+		EMSG("STGEN configuration failed, error code: %#llx",
+		     (unsigned long long)stgen_conf_args.a0);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void stm32_stgen_pm_suspend(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	/*
+	 * Disable STGEN counter. At this point, it should be stopped by
+	 * software.
+	 */
+	io_clrbits32(stgen_d.base + STGENC_CNTCR, STGENC_CNTCR_EN);
+
+	/* Save current time from the RTC */
+	res = rtc_get_time(stgen_d.calendar);
+	if (res)
+		panic("Could not get RTC calendar at suspend");
+
+	/* Save current counter value */
+	reg_pair_from_64(stm32_stgen_get_counter_value(),
+			 &stgen_d.cnt_h, &stgen_d.cnt_l);
+
+	/* Re-enable STGEN as generic timer can be used later */
+	io_setbits32(stgen_d.base + STGENC_CNTCR, STGENC_CNTCR_EN);
+}
+
+static void stm32_stgen_pm_resume(void)
+{
+	uint64_t counter_val = reg_pair_to_64(stgen_d.cnt_h, stgen_d.cnt_l);
+	unsigned long clock_src_rate = clk_get_rate(stgen_d.stgen_clock);
+	struct optee_rtc_time cur_calendar = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint64_t nb_pm_count_ticks = 0;
+
+	/* STGEN clocks are already restored by the clock driver */
+	io_clrbits32(stgen_d.base + STGENC_CNTCR, STGENC_CNTCR_EN);
+
+	/* Read the current time from the RTC to update system counter */
+	res = rtc_get_time(&cur_calendar);
+	if (res)
+		panic("Could not get RTC calendar at resume");
+
+	nb_pm_count_ticks = stm32_rtc_diff_calendar_tick(&cur_calendar,
+							 stgen_d.calendar,
+							 clock_src_rate);
+
+	DMSG("Time spent in low-power: %"PRIu64"ms",
+	     (nb_pm_count_ticks * 1000) / clock_src_rate);
```

Comment:
For a better accuracy of the restored time, the trace message should better be printed when exiting the function, after STGEN counting is enabled back.